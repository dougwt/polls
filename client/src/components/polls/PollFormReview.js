import React from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { Card, Input, Button, Row, Icon } from 'react-materialize';
import * as actions from '../../actions';

export const PollFormReview = ({ onCancel, formValues, submitPoll, history }) => {
  return (
    <div>
      <h5>Please confirm your poll</h5>

      {/* {console.log(formValues)} */}

      <Row>
    		<Card className='darken-1' title={formValues.question}>

          {/* TODO: add spacing here via CSS */}

          <div className="choices">
            {renderChoices(formValues)}
          </div>

          <Row className="center-align">
            <Button
              className="teal"
              disabled
            >
              Vote
            </Button>
          </Row>

        </Card>
      </Row>

      <Row>
        <Button
          node="a"
          className="red btn-back btn-flat white-text"
          waves="light"
          onClick={onCancel}
        >
          Back
          <Icon left>keyboard_arrow_left</Icon>
        </Button>

        <Button
          className="teal btn-create btn-flat right white-text"
          waves="light"
          onClick={() => submitPoll(formValues, history)}
        >
          Create
          <Icon right>create</Icon>
        </Button>
      </Row>
    </div>
  )
};

function renderChoices(values) {
  let choices = [];

  for (let choice = 1; choice <= values.choices.length; choice++) {
    let field = `choice_${choice}`;
    if(values[field]) {
      choices.push(
        <Row key={choice}>
          <Input name="choices" type="radio" value={choice.toString()} label={values[field]} />
        </Row>
      )
    }
  }

  return choices;
};

function mapStateToProps(state) {
  return {
    formValues: state.form.pollNewForm.values
  };
}

export default connect(mapStateToProps, actions)(withRouter(PollFormReview));
