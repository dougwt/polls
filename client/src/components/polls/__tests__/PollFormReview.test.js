import React from 'react';
import { shallow } from 'enzyme';
import { PollFormReview } from '../PollFormReview';

describe('PollFormReview', () => {
  let props;
  let wrapper;

  beforeEach(() => {
    props = {
      error: null,
      formValues: {
        choices: [0, 1, undefined, undefined, undefined, undefined],
        question: 'Who is your favorite Starfleet captain?'
      },
      history: {},
      onCancel: jest.fn(),
      onSave: jest.fn(),
      waiting: false
    };

    wrapper = shallow(<PollFormReview {...props} />);
  });

  it('renders properly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('shows a PollDetail', () => {
    expect(wrapper.find('PollDetail').length).toBe(1);
  });

  it('shows a Back button', () => {
    expect(wrapper.find('Button.btn-back').length).toBe(1);
  });

  it('shows a Next button', () => {
    expect(wrapper.find('Button.btn-next').length).toBe(1);
  });

  it('does not show a loading spinner initially', () => {
    expect(wrapper.find('ProgressBar').length).toEqual(0);
  });

  it('does not show an error message initially', () => {
    expect(wrapper.find('Row.error').length).toEqual(0);
  });

  describe('when the Back button is clicked', () => {
    it('executes the `onCancel` callback in props', () => {
      wrapper.find('Button.btn-back').simulate('click');
      expect(props.onCancel).toHaveBeenCalled();
    });
  });

  describe('when the Next button is clicked', () => {
    it('executes the `onSave` callback in props', () => {
      wrapper.find('Button.btn-next').simulate('click');
      expect(props.onSave).toHaveBeenCalled();
    });

    describe('when the save operation is waiting', () => {
      beforeEach(() => {
        wrapper = shallow(<PollFormReview {...props} waiting={true} />);
      });

      it('can show a loading spinner during save', () => {
        expect(wrapper.find('ProgressBar').length).toEqual(1);
      });
    });

    describe('when the save operation fails', () => {
      beforeEach(() => {
        wrapper = shallow(
          <PollFormReview
            {...props}
            error={{ response: { data: { error: 'owch!' } } }}
          />
        );
      });

      it('can show an error message', () => {
        expect(wrapper.find('Row.error').length).toEqual(1);
      });
    });
  });
});
