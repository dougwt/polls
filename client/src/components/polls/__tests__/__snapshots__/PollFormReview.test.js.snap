// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PollFormReview renders properly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Unknown
    error={null}
    formValues={
        Object {
            "choices": Array [
              0,
              1,
              undefined,
              undefined,
              undefined,
              undefined,
            ],
            "question": "Who is your favorite Starfleet captain?",
          }
    }
    onCancel={[Function]}
    onSave={[Function]}
    waiting={false}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h5>
          Please confirm your poll
</h5>,
        <PollDetail
          disabled={true}
          formValues={
                    Object {
                              "choices": Array [
                                0,
                                1,
                                undefined,
                                undefined,
                                undefined,
                                undefined,
                              ],
                              "question": "Who is your favorite Starfleet captain?",
                            }
          }
/>,
        <Row>
           
</Row>,
        <Row>
          <Button
                    className="red btn-back white-text"
                    node="a"
                    onClick={[Function]}
                    waves="light"
          >
                    Back
                    <Icon
                              left={true}
                    >
                              keyboard_arrow_left
                    </Icon>
          </Button>
          <Button
                    className="teal btn-next right white-text"
                    node="button"
                    onClick={[Function]}
                    waves="light"
          >
                    Create
                    <Icon
                              right={true}
                    >
                              create
                    </Icon>
          </Button>
</Row>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Please confirm your poll",
        },
        "ref": null,
        "rendered": "Please confirm your poll",
        "type": "h5",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "disabled": true,
          "formValues": Object {
            "choices": Array [
              0,
              1,
              undefined,
              undefined,
              undefined,
              undefined,
            ],
            "question": "Who is your favorite Starfleet captain?",
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": " ",
        },
        "ref": null,
        "rendered": " ",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <Button
              className="red btn-back white-text"
              node="a"
              onClick={[Function]}
              waves="light"
>
              Back
              <Icon
                            left={true}
              >
                            keyboard_arrow_left
              </Icon>
</Button>,
            <Button
              className="teal btn-next right white-text"
              node="button"
              onClick={[Function]}
              waves="light"
>
              Create
              <Icon
                            right={true}
              >
                            create
              </Icon>
</Button>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": Array [
                "Back",
                <Icon
                  left={true}
>
                  keyboard_arrow_left
</Icon>,
              ],
              "className": "red btn-back white-text",
              "node": "a",
              "onClick": [Function],
              "waves": "light",
            },
            "ref": null,
            "rendered": Array [
              "Back",
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "children": "keyboard_arrow_left",
                  "left": true,
                },
                "ref": null,
                "rendered": "keyboard_arrow_left",
                "type": [Function],
              },
            ],
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": Array [
                "Create",
                <Icon
                  right={true}
>
                  create
</Icon>,
              ],
              "className": "teal btn-next right white-text",
              "node": "button",
              "onClick": [Function],
              "waves": "light",
            },
            "ref": null,
            "rendered": Array [
              "Create",
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "children": "create",
                  "right": true,
                },
                "ref": null,
                "rendered": "create",
                "type": [Function],
              },
            ],
            "type": [Function],
          },
        ],
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h5>
            Please confirm your poll
</h5>,
          <PollDetail
            disabled={true}
            formValues={
                        Object {
                                    "choices": Array [
                                      0,
                                      1,
                                      undefined,
                                      undefined,
                                      undefined,
                                      undefined,
                                    ],
                                    "question": "Who is your favorite Starfleet captain?",
                                  }
            }
/>,
          <Row>
             
</Row>,
          <Row>
            <Button
                        className="red btn-back white-text"
                        node="a"
                        onClick={[Function]}
                        waves="light"
            >
                        Back
                        <Icon
                                    left={true}
                        >
                                    keyboard_arrow_left
                        </Icon>
            </Button>
            <Button
                        className="teal btn-next right white-text"
                        node="button"
                        onClick={[Function]}
                        waves="light"
            >
                        Create
                        <Icon
                                    right={true}
                        >
                                    create
                        </Icon>
            </Button>
</Row>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Please confirm your poll",
          },
          "ref": null,
          "rendered": "Please confirm your poll",
          "type": "h5",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "disabled": true,
            "formValues": Object {
              "choices": Array [
                0,
                1,
                undefined,
                undefined,
                undefined,
                undefined,
              ],
              "question": "Who is your favorite Starfleet captain?",
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": " ",
          },
          "ref": null,
          "rendered": " ",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <Button
                className="red btn-back white-text"
                node="a"
                onClick={[Function]}
                waves="light"
>
                Back
                <Icon
                                left={true}
                >
                                keyboard_arrow_left
                </Icon>
</Button>,
              <Button
                className="teal btn-next right white-text"
                node="button"
                onClick={[Function]}
                waves="light"
>
                Create
                <Icon
                                right={true}
                >
                                create
                </Icon>
</Button>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": Array [
                  "Back",
                  <Icon
                    left={true}
>
                    keyboard_arrow_left
</Icon>,
                ],
                "className": "red btn-back white-text",
                "node": "a",
                "onClick": [Function],
                "waves": "light",
              },
              "ref": null,
              "rendered": Array [
                "Back",
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "children": "keyboard_arrow_left",
                    "left": true,
                  },
                  "ref": null,
                  "rendered": "keyboard_arrow_left",
                  "type": [Function],
                },
              ],
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": Array [
                  "Create",
                  <Icon
                    right={true}
>
                    create
</Icon>,
                ],
                "className": "teal btn-next right white-text",
                "node": "button",
                "onClick": [Function],
                "waves": "light",
              },
              "ref": null,
              "rendered": Array [
                "Create",
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "children": "create",
                    "right": true,
                  },
                  "ref": null,
                  "rendered": "create",
                  "type": [Function],
                },
              ],
              "type": [Function],
            },
          ],
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
